File: server/publish.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

// publicPublisher mirrors live edits straight into the same R2 bucket the
// static viewer already reads from (see tools/buildstatic and
// .github/workflows/deploy-r2.yml), in the exact key shape buildstatic
// produces -- sites.json at the bucket root, media at media/<id>/<filename>
// -- so a save is visible on the public site within seconds instead of
// waiting on a GitHub Actions rebuild (Phase 8, github issue #9).
//
// This is a best-effort mirror, not the source of truth: R2_LIVE_* (see
// r2SiteStore) remains authoritative. A publish failure is logged, not
// surfaced to the API caller -- the edit itself already succeeded by the
// time this runs.
//
// Configured by a second set of R2 credentials (R2_PUBLIC_*) scoped to the
// public bucket, separate from R2_LIVE_* -- deliberately optional: with
// them unset, newPublicPublisherFromEnv returns nil and every call below is
// a no-op, so this only activates once those credentials are provisioned
// and wired into Cloud Run (see README).
type publicPublisher struct {
	client *s3.Client
	bucket string
}

func newPublicPublisherFromEnv() *publicPublisher {
	accountID := os.Getenv("R2_ACCOUNT_ID")
	accessKeyID := os.Getenv("R2_PUBLIC_ACCESS_KEY_ID")
	secretAccessKey := os.Getenv("R2_PUBLIC_SECRET_ACCESS_KEY")
	bucket := os.Getenv("R2_PUBLIC_BUCKET_NAME")
	if accountID == "" || accessKeyID == "" || secretAccessKey == "" || bucket == "" {
		log.Printf("[publish] R2_PUBLIC_* not fully configured -- edits will only reach the git+Actions static deploy, not publish instantly")
		return nil
	}
	client := s3.New(s3.Options{
		Region:       "auto",
		BaseEndpoint: aws.String(fmt.Sprintf("https://%s.r2.cloudflarestorage.com", accountID)),
		Credentials:  credentials.NewStaticCredentialsProvider(accessKeyID, secretAccessKey, ""),
	})
	return &publicPublisher{client: client, bucket: bucket}
}

// publishSites overwrites sites.json with the full current site list --
// buildstatic writes this same file in this same shape, so the static
// viewer needs no changes to read whichever one is newer.
func (p *publicPublisher) publishSites(ctx context.Context, sites []Site) {
	if p == nil {
		return
	}
	data, err := json.Marshal(sites)
	if err != nil {
		log.Printf("[publish] marshaling sites.json: %v", err)
		return
	}
	_, err = p.client.PutObject(ctx, &s3.PutObjectInput{
		Bucket:      aws.String(p.bucket),
		Key:         aws.String("sites.json"),
		Body:        bytes.NewReader(data),
		ContentType: aws.String("application/json; charset=utf-8"),
	})
	if err != nil {
		log.Printf("[publish] writing sites.json: %v", err)
	}
}

// publishMedia mirrors one uploaded file to media/<id>/<filename>, matching
// buildstatic's copyMedia layout (flat under media/, not nested under
// sites/ the way the live-data bucket and local disk are).
func (p *publicPublisher) publishMedia(ctx context.Context, id, filename string, data []byte) {
	if p == nil {
		return
	}
	_, err := p.client.PutObject(ctx, &s3.PutObjectInput{
		Bucket:      aws.String(p.bucket),
		Key:         aws.String("media/" + id + "/" + filename),
		Body:        bytes.NewReader(data),
		ContentType: aws.String(mediaContentType(filename)),
	})
	if err != nil {
		log.Printf("[publish] writing media/%s/%s: %v", id, filename, err)
	}
}

func (p *publicPublisher) deleteMedia(ctx context.Context, id, filename string) {
	if p == nil {
		return
	}
	_, err := p.client.DeleteObject(ctx, &s3.DeleteObjectInput{
		Bucket: aws.String(p.bucket),
		Key:    aws.String("media/" + id + "/" + filename),
	})
	if err != nil {
		log.Printf("[publish] deleting media/%s/%s: %v", id, filename, err)
	}
}

// deleteSiteMedia removes every object under media/<id>/ -- called when a
// site is deleted, since sites.json's own rewrite (via publishSites) only
// covers the site list, not its now-orphaned media files.
func (p *publicPublisher) deleteSiteMedia(ctx context.Context, id string) {
	if p == nil {
		return
	}
	prefix := "media/" + id + "/"
	var continuationToken *string
	for {
		out, err := p.client.ListObjectsV2(ctx, &s3.ListObjectsV2Input{
			Bucket:            aws.String(p.bucket),
			Prefix:            aws.String(prefix),
			ContinuationToken: continuationToken,
		})
		if err != nil {
			log.Printf("[publish] listing %s: %v", prefix, err)
			return
		}
		for _, obj := range out.Contents {
			if _, err := p.client.DeleteObject(ctx, &s3.DeleteObjectInput{
				Bucket: aws.String(p.bucket),
				Key:    obj.Key,
			}); err != nil {
				log.Printf("[publish] deleting %s: %v", aws.ToString(obj.Key), err)
			}
		}
		if out.IsTruncated == nil || !*out.IsTruncated {
			return
		}
		continuationToken = out.NextContinuationToken
	}
}

File: server/src/publish.ts
// Mirrors live edits into the PUBLIC_SITE store (sites.json at the root,
// media at media/<id>/<filename>) so the public read-only viewer's
// GET /sites.json reflects an edit within seconds, without touching
// LIVE_DATA (and its per-site YAML parse) on every anonymous page view.
// Ports server/publish.go's publicPublisher -- but with no optional
// credentials to configure, since both stores are just ObjectStore
// instances here. See github issue #18.
//
// Best-effort like the Go version: source of truth is LIVE_DATA (see
// store.ts); a publish failure here would only delay the public mirror, not
// the edit itself, so callers can treat these as fire-and-forget.

import { type Site } from "./siteyaml";
import { getAllSites, mediaContentType } from "./store";
import type { ObjectStore } from "./objectstore";

const textDecoder = new TextDecoder();

export async function publishAllSites(live: ObjectStore, publicStore: ObjectStore): Promise<void> {
  const sites = await getAllSites(live);
  sites.sort((a, b) => a.id.localeCompare(b.id));
  await publishSites(publicStore, sites);
}

async function publishSites(store: ObjectStore, sites: Site[]): Promise<void> {
  await store.put("sites.json", JSON.stringify(sites), "application/json; charset=utf-8");
}

async function readPublishedSites(store: ObjectStore): Promise<Site[] | null> {
  const obj = await store.get("sites.json");
  if (!obj) return null;
  try {
    return JSON.parse(textDecoder.decode(obj.data)) as Site[];
  } catch {
    return null; // corrupt snapshot -- treat the same as missing
  }
}

// Patches one created/updated site into the published sites.json instead
// of rebuilding it from every site in LIVE_DATA the way publishAllSites
// does. With dozens of sites, that full re-fetch (even parallelized -- see
// getAllSites) was the dominant cost of a save: ~2.5s of it measured
// against 85 sites, just to regenerate a snapshot where only one entry
// actually changed. Falls back to publishAllSites if the current mirror
// is missing or unreadable, so a broken/absent snapshot still self-heals
// rather than staying broken.
export async function publishSiteChange(live: ObjectStore, publicStore: ObjectStore, site: Site): Promise<void> {
  const current = await readPublishedSites(publicStore);
  if (current == null) {
    await publishAllSites(live, publicStore);
    return;
  }
  const next = current.filter((s) => s.id !== site.id);
  next.push(site);
  next.sort((a, b) => a.id.localeCompare(b.id));
  await publishSites(publicStore, next);
}

// Same idea as publishSiteChange, for a deleted site.
export async function publishSiteRemoval(live: ObjectStore, publicStore: ObjectStore, id: string): Promise<void> {
  const current = await readPublishedSites(publicStore);
  if (current == null) {
    await publishAllSites(live, publicStore);
    return;
  }
  await publishSites(
    publicStore,
    current.filter((s) => s.id !== id),
  );
}

// Mirrors one uploaded file to media/<id>/<filename> (flat, not nested
// under sites/ the way LIVE_DATA is).
export async function publishMedia(store: ObjectStore, id: string, filename: string, data: ArrayBuffer | Uint8Array): Promise<void> {
  await store.put(`media/${id}/${filename}`, data, mediaContentType(filename));
}

export async function publishDeleteMedia(store: ObjectStore, id: string, filename: string): Promise<void> {
  await store.delete(`media/${id}/${filename}`);
}

// Removes every object under media/<id>/ -- called when a site is deleted,
// since the sites.json rewrite above only covers the site list, not its
// now-orphaned media files.
export async function publishDeleteSiteMedia(store: ObjectStore, id: string): Promise<void> {
  await store.deleteMany(await store.list(`media/${id}/`));
}
